### PR TITLE
Implement `visit_Entry()`

### LINKED ISSUE #1836: [SciPy][cdflib]: visit_Entry() not implemented
```fortran
SUBROUTINE dzror()
print *, "in dzror"
ENTRY dstzr()
    print *, "in dztzr"
RETURN
END

program main
    print *, "Call dzror"
    call dzror()
    print *, "Call dstzr"
    call dstzr()
end program

```

> GFortran

```console
% gfortran a.f90 && ./a.out
 Call dzror
 in dzror
 in dztzr
 Call dstzr
 in dztzr
```

> LFortran

```console
% lfortran a.f90
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
LCompilersException: visit_Entry() not implemented
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1836.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
`integration_tests/entry_02.f90` shows very weird behaviour, I am not sure why this happens: 

> entry_02

```fortran
subroutine x(dummy)
   real :: dummy
   print *, "Printed using subroutine call: ", dummy
   if (abs(dummy - 10.0) > 1e-7) error stop
   dummy = 5.0
 entry y(dummy)
   print *, "Printed using entry statement: ", dummy
   print *, abs(dummy - 5.0)
   if (abs(dummy - 5.0) > 1e-7) error stop
   return
end subroutine

program entry_02
   real :: dummy
   dummy = 10.0
   call x(dummy)
   dummy = 5.0
   call y(dummy)
end program
```


> LFortran

```console
% lfortran ./integration_tests/entry_02.f90 --implicit-typing --implicit-interface
Printed using subroutine call:  1.00000000e+01
Printed using entry statement:  5.00000000e+00
0.00000000e+00
Printed using entry statement:  5.00000000e+00
0.00000000e+00
ERROR STOP
```

**Comment by certik** [comment]
Very good. We just need to debug it. Print the ASR that LLVM sees, as well as LLVM code and let's figure out where the problem is.

**Comment by Pranavchiku** [comment]
This is tricky, I know why this happens, trying to resolve this.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 1]
```diff
@@ -82,6 +82,64 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+    void copy_stmts(size_t n_body, AST::stmt_t **m_body) {
+        std::vector<std::string> entry_points;
+        for (size_t i=0; i<n_body; i++) {
+            if (m_body[i]->type == AST::stmtType::Entry) {
+                AST::Entry_t* entry = (AST::Entry_t*)m_body[i];
+                entry_points.push_back(std::string(entry->m_name));
+            } else {
+                for (auto sym_name: entry_points) {
+                    entry_point_mapping[sym_name].push_back(m_body[i]);
+                }
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:94]: So if `entry_points` is empty (the vast majority of cases), then this does nothing?

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 2]
```diff
@@ -1603,6 +1661,90 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
                     del_syms.p, del_syms.size()));
     }
 
+    void visit_Entry(const AST::Entry_t &x) {
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:1664]: Let's make a note that entry points are present. Say:
```
entry_points_present = true;
```
And make sure you set it to false at the beginning of all functions.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 3]
```diff
@@ -1629,6 +1771,10 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         body.reserve(al, x.n_body);
         format_statements.clear();
         transform_stmts(body, x.n_body, x.m_body);
+        copy_stmts(x.n_body, x.m_body);
+        handle_active_entry_points();
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:1775]: Now let's do this, to ensure fast execution in the default case (no entry points):
```suggestion
        if (entry_points_present) {
            copy_stmts(x.n_body, x.m_body);
            handle_active_entry_points();
        }
```
```

**Comment by certik** [comment]
Very good. Does this handle multiple entry points? If so, then let's add tests for 2 and 3 entry points.

File: src/lfortran/semantics/ast_body_visitor.cpp (hunk 1 above)
Comment (Pranavchiku) [reply src/lfortran/semantics/ast_body_visitor.cpp:94]: Correct

File: src/lfortran/semantics/ast_body_visitor.cpp (hunk 2 above)
Comment (Pranavchiku) [reply src/lfortran/semantics/ast_body_visitor.cpp:1664]: Yes, perfect.

**Review by certik — APPROVED** [review]
I think this looks great now! Thanks!

**Pull request merged**